package dev.melodify.uranophilelab.activities

import android.content.Context
import android.content.DialogInterface
import android.content.SharedPreferences
import android.os.Bundle
import android.view.View
import android.widget.RadioGroup
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import dev.melodify.uranophilelab.BaseApplicationClass
import dev.melodify.uranophilelab.R
import dev.melodify.uranophilelab.databinding.ActivitySettingsBinding
import dev.melodify.uranophilelab.utils.SharedPreferenceManager
import dev.melodify.uranophilelab.utils.customview.MaterialCustomSwitch.OnCheckChangeListener
import dev.melodify.uranophilelab.utils.MiniPlayerHelper
import androidx.core.content.edit

class SettingsActivity : AppCompatActivity() {
    var binding: ActivitySettingsBinding? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySettingsBinding.inflate(layoutInflater)
        setContentView(binding!!.getRoot())
        MiniPlayerHelper.initMiniPlayer(this)
        val settingsSharedPrefManager = SettingsSharedPrefManager(this)
        val sharedPreferenceManager: SharedPreferenceManager =
            SharedPreferenceManager.getInstance(this)

        binding!!.downloadOverCellular.setOnCheckChangeListener(object : OnCheckChangeListener {
            override fun onCheckChanged(isChecked: Boolean) {
                settingsSharedPrefManager.downloadOverCellular = isChecked
            }
        })
        binding!!.highQualityTrack.setOnCheckChangeListener(object : OnCheckChangeListener {
            override fun onCheckChanged(isChecked: Boolean) {
                settingsSharedPrefManager.highQualityTrack = isChecked
            }
        })
        binding!!.storeInCache.setOnCheckChangeListener(object : OnCheckChangeListener {
            override fun onCheckChanged(isChecked: Boolean) {
                settingsSharedPrefManager.storeInCache = isChecked
            }
        })
        binding!!.explicit.setOnCheckChangeListener(object : OnCheckChangeListener {
            override fun onCheckChanged(isChecked: Boolean) {
                settingsSharedPrefManager.explicit = isChecked
            }
        })
        binding!!.playInBackground?.setOnCheckChangeListener(object : OnCheckChangeListener {
            override fun onCheckChanged(isChecked: Boolean) {
                settingsSharedPrefManager.playInBackground = isChecked
            }
        })

        binding!!.downloadOverCellular.setChecked(settingsSharedPrefManager.downloadOverCellular)
        binding!!.highQualityTrack.setChecked(settingsSharedPrefManager.highQualityTrack)
        binding!!.storeInCache.setChecked(settingsSharedPrefManager.storeInCache)
        binding!!.explicit.setChecked(settingsSharedPrefManager.explicit)
        binding!!.playInBackground?.setChecked(settingsSharedPrefManager.playInBackground)

        binding!!.themeChipGroup.setOnCheckedChangeListener { _: RadioGroup?, checkedId: Int ->
            settingsSharedPrefManager.theme =
                if (checkedId == R.id.dark) "dark" else if (checkedId == R.id.light) "light" else "system"
            BaseApplicationClass.updateTheme()
        }

        binding!!.clearCache.setOnClickListener {
            AlertDialog.Builder(this)
                .setTitle("Clear Cache")
                .setMessage("Are you sure you want to clear the cache?")
                .setPositiveButton(
                    "Yes"
                ) { _: DialogInterface?, _: Int ->
                    sharedPreferenceManager.clearOldPrefsAsync(
                        this@SettingsActivity,
                        null
                    )
                }
                .setNegativeButton("No", null)
                .show()
        }

        binding!!.themeChipGroup.check(if (settingsSharedPrefManager.theme == "dark") R.id.dark else if (settingsSharedPrefManager.theme == "light") R.id.light else R.id.system)
    }

    override fun onResume() {
        super.onResume()
        MiniPlayerHelper.onActivityResume(this)
    }

    override fun onPause() {
        super.onPause()
        MiniPlayerHelper.onActivityPause(this)
    }

    fun backPress(view: View?) {
        finish()
    }

    class SettingsSharedPrefManager(context: Context) {
        var sharedPreferences: SharedPreferences = context.getSharedPreferences("settings", MODE_PRIVATE)

        var downloadOverCellular: Boolean
            get() = sharedPreferences.getBoolean("download_over_cellular", true)
            set(value) {
                sharedPreferences.edit { putBoolean("download_over_cellular", value) }
            }

        var highQualityTrack: Boolean
            get() = sharedPreferences.getBoolean("high_quality_track", true)
            set(value) {
                sharedPreferences.edit { putBoolean("high_quality_track", value) }
            }

        var storeInCache: Boolean
            get() = sharedPreferences.getBoolean("store_in_cache", true)
            set(value) {
                sharedPreferences.edit { putBoolean("store_in_cache", value) }
            }

        var explicit: Boolean
            get() = sharedPreferences.getBoolean("explicit", true)
            set(value) {
                sharedPreferences.edit { putBoolean("explicit", value) }
            }

        var playInBackground: Boolean
            get() = sharedPreferences.getBoolean("play_in_background", true)
            set(value) {
                sharedPreferences.edit { putBoolean("play_in_background", value) }
            }

        var theme: String?
            get() = sharedPreferences.getString("theme", "system")
            set(theme) {
                sharedPreferences.edit { putString("theme", theme) }
            }
    }
}
